src/pdn/src/PdnGen.cc

```diff
@@ -225,14 +251,629 @@ PdnGen::addGlobalConnect(dbBox* region, const char* instPattern, const char* pin
+        } else {
+          component->replaceShape(shape.get(), {new_shape});
+        }
+      }
+    }
+  }
+  debugPrint(logger_, utl::PDN, "Make", 2, "Trim shapes - end");
+}
+
+VoltageDomain* PdnGen::getCoreDomain()
+{
+  if (core_domain_ == nullptr) {
+    setCoreDomain(nullptr, nullptr, {});
+  }
+
+  return core_domain_.get();
+}
+
+std::vector<VoltageDomain*> PdnGen::getDomains()
+{
+  std::vector<VoltageDomain*> domains;
+  domains.push_back(getCoreDomain());
+
+  for (const auto& domain : domains_) {
+    domains.push_back(domain.get());
+  }
+  return domains;
+}
+
+std::vector<VoltageDomain*> PdnGen::getConstDomains() const
+{
+  std::vector<VoltageDomain*> domains;
+  if (core_domain_ != nullptr) {
+    domains.push_back(core_domain_.get());
+  }
+
+  for (const auto& domain : domains_) {
+    domains.push_back(domain.get());
+  }
+  return domains;
+}
+
+VoltageDomain* PdnGen::findDomain(const std::string& name)
+{
+  auto domains = getDomains();
+  if (name.empty()) {
+    if (domains.empty()) {
+      return nullptr;
+    }
+
+    return domains.back();
+  }
+
+  for (const auto& domain : domains) {
+    if (domain->getName() == name) {
+      return domain;
+    }
+  }
+
+  return nullptr;
+}
+
+void PdnGen::setCoreDomain(odb::dbNet* power,
+                           odb::dbNet* ground,
+                           const std::vector<odb::dbNet*>& secondary)
+{
+  auto* block = db_->getChip()->getBlock();
+  if (core_domain_ != nullptr) {
+    logger_->warn(utl::PDN, 183, "Replacing existing core voltage domain.");
+  }
+  core_domain_ = std::make_unique<VoltageDomain>(
+      this, block, power, ground, secondary, logger_);
+}
+
+void PdnGen::makeRegionVoltageDomain(
+    const std::string& name,
+    odb::dbNet* power,
+    odb::dbNet* ground,
+    const std::vector<odb::dbNet*>& secondary_nets,
+    odb::dbRegion* region)
+{
+  for (const auto& domain : domains_) {
+    if (domain->getName() == name) {
+      logger_->error(utl::PDN,
+                     184,
+                     "Cannot have region voltage domain with the same name "
+                     "already exists: {}",
+                     name);
+    }
+  }
+  auto* block = db_->getChip()->getBlock();
+  domains_.push_back(std::make_unique<VoltageDomain>(
+      this, name, block, power, ground, secondary_nets, region, logger_));
+}
+
+std::vector<Grid*> PdnGen::getGrids() const
+{
+  std::vector<Grid*> grids;
+  if (core_domain_ != nullptr) {
+    for (const auto& grid : core_domain_->getGrids()) {
+      grids.push_back(grid.get());
+    }
+  }
+  for (const auto& domain : domains_) {
+    for (const auto& grid : domain->getGrids()) {
+      grids.push_back(grid.get());
+    }
+  }
+
+  return grids;
+}
+
+std::vector<Grid*> PdnGen::findGrid(const std::string& name) const
+{
+  std::vector<Grid*> found_grids;
+  auto grids = getGrids();
+
+  if (name.empty()) {
+    if (grids.empty()) {
+      return {};
+    }
+
+    return findGrid(grids.back()->getName());
```

Comment:
Would this produce something other than {grids.back()} ?

Reply:
To emulate the behavior of the tcl version, I need to get the full list of grids that share that name, not just the very last one. This is because the instances end up generating different grids for each instance while sharing the same name.